Code/Tools/ProjectManager/Source/GemCatalog/GemCatalogScreen.cpp

```diff
@@ -144,11 +149,80 @@ namespace O3DE::ProjectManager
                 {
                     m_gemModel->AddGem(gemInfoResult.GetValue<GemInfo>());
                     m_gemModel->UpdateGemDependencies();
+                    m_proxyModel->sort(/*column=*/0);
                 }
             }
         }
     }
 
+    void GemCatalogScreen::Refresh(const QString& projectPath)
+    {
+        QHash<QString, GemInfo> gemInfoHash;
+
+        // create a hash with the gem name as key
+        AZ::Outcome<QVector<GemInfo>, AZStd::string> allGemInfosResult = PythonBindingsInterface::Get()->GetAllGemInfos(projectPath);
+        if (allGemInfosResult.IsSuccess())
+        {
+            QVector<GemInfo> gemInfos = allGemInfosResult.GetValue();
+            for (const GemInfo& gemInfo : gemInfos)
+            {
+                gemInfoHash.insert(gemInfo.m_name, gemInfo);
+            }
+        }
+
+        // add all the gem repos into the hash
+        AZ::Outcome<QVector<GemInfo>, AZStd::string> allRepoGemInfosResult = PythonBindingsInterface::Get()->GetAllGemRepoGemsInfos();
+        if (allRepoGemInfosResult.IsSuccess())
+        {
+            const QVector<GemInfo> allRepoGemInfos = allRepoGemInfosResult.GetValue();
+            for (const GemInfo& gemInfo : allRepoGemInfos)
+            {
+                if (!gemInfoHash.contains(gemInfo.m_name))
+                {
+                    gemInfoHash.insert(gemInfo.m_name, gemInfo);
+                }
+            }
+        }
+
+        // remove gems from the model that no longer exist in the hash and are not project dependencies
+        int i = 0;
+        while (i < m_gemModel->rowCount())
+        {
+            QModelIndex index = m_gemModel->index(i,0);
+            QString gemName = m_gemModel->GetName(index);
+            const bool gemFound = gemInfoHash.contains(gemName);
+            if (!gemFound && !m_gemModel->IsAdded(index) && !m_gemModel->IsAddedDependency(index))
+            {
+                m_gemModel->removeRow(i);
+            }
+            else
+            {
+                if (!gemFound && (m_gemModel->IsAdded(index) || m_gemModel->IsAddedDependency(index)))
+                {
+                    const QString error = tr("Gem %1 was removed or unregistered, but is still used by the project.").arg(gemName);
+                    AZ_Warning("Project Manager", false, error.toUtf8().constData());
+                    QMessageBox::warning(this, tr("Gem not found"), error.toUtf8().constData());
```

Comment:
I feel like this needs more of a call to action for the user. Or stricter enforcement because couldn't this cause serious problems for their project? Maybe it could ask and recommend that they remove it, with a button that does so?

Reply:
That's a good suggestion, but it's an edge case I haven't even been able to simulate, so lets leave it for another PR.